pkg/autodiscovery/listeners/kube_endpoints.go

```diff
@@ -0,0 +1,336 @@
+	if first.ResourceVersion == second.ResourceVersion {
+		return false
+	}
+	// AD annotations on the corresponding service
+	if l.isEndpointsAnnotated(first) != l.isEndpointsAnnotated(second) {
+		return true
+	}
+	// Subsets
+	return subsetsDiffer(first, second)
+}
+
+// subsetsDiffer detects if two Endpoints have different subsets.
+// The function is separated from endpointsDiffer to facilitate testing.
+func subsetsDiffer(first, second *v1.Endpoints) bool {
+	// Subsets length
+	if len(first.Subsets) != len(second.Subsets) {
+		return true
+	}
+	// Addresses and Ports
+	for i := range first.Subsets {
+		if len(first.Subsets[i].Addresses) != len(second.Subsets[i].Addresses) {
+			return true
+		}
+		if len(first.Subsets[i].Ports) != len(second.Subsets[i].Ports) {
+			return true
+		}
+		// Addresses
+		for j := range first.Subsets[i].Addresses {
+			if first.Subsets[i].Addresses[j].IP != second.Subsets[i].Addresses[j].IP {
+				return true
+			}
+			if first.Subsets[i].Addresses[j].Hostname != second.Subsets[i].Addresses[j].Hostname {
+				return true
+			}
+		}
+		// Ports
+		for j := range first.Subsets[i].Ports {
+			if first.Subsets[i].Ports[j].Port != second.Subsets[i].Ports[j].Port {
+				return true
+			}
+			if first.Subsets[i].Ports[j].Name != second.Subsets[i].Ports[j].Name {
+				return true
+			}
+		}
+	}
+	// No relevant change
+	return false
+}
+
+// isEndpointsAnnotated looks for the corresponding service of a kubernetes endpoints object
+// and returns true if the service has endpoints annotations, otherwise returns false.
+func (l *KubeEndpointsListener) isEndpointsAnnotated(kep *v1.Endpoints) bool {
+	ksvc, err := l.serviceLister.Services(kep.Namespace).Get(kep.Name)
+	if err != nil {
+		log.Debugf("Cannot get Kubernetes service: %s", err)
+	}
+	if ksvc != nil {
+		if _, found := ksvc.Annotations[kubeEndpointsAnnotationFormat]; found {
+			return true
+		}
+	}
+	return false
+}
+
+func (l *KubeEndpointsListener) createService(kep *v1.Endpoints, firstRun bool) {
+	if kep == nil {
+		return
+	}
+	if !l.isEndpointsAnnotated(kep) {
+		// Ignore endpoints with no AD annotation on their corresponding service
+		return
+	}
+
+	eps := processEndpoints(kep, firstRun)
+
+	l.m.Lock()
+	l.endpoints[kep.UID] = eps
+	l.m.Unlock()
+
+	for _, ep := range eps {
+		l.newService <- ep
+	}
+}
+
+// processEndpoints parses a kubernetes Endpoints object
+// and returns a slice of KubeEndpointService per endpoint
+func processEndpoints(kep *v1.Endpoints, firstRun bool) []*KubeEndpointService {
+	var eps []*KubeEndpointService
+	for i := range kep.Subsets {
+		ports := []ContainerPort{}
+		// Ports
+		for _, port := range kep.Subsets[i].Ports {
+			ports = append(ports, ContainerPort{int(port.Port), port.Name})
+		}
+		// Hosts
+		for _, host := range kep.Subsets[i].Addresses {
+			// create a separate AD service per host
+			ep := &KubeEndpointService{
+				entity:       apiserver.EntityForEndpoints(kep.Namespace, kep.Name, host.IP),
```

Comment:
It's because we generate a separate AD service per IP.
For example, if the Endpoints object has 3 IPs we need to generate 3 config templates with the same check config but with different entities and AD IDs.
If we use the Endpoints UID we will generate 3 templates with the same entity values. Thus, we use the IP to make unique entities. The logic of generating these templates is in the config provider, we have a function with this signature

```
// generateConfigs creates a config template for each Endpoints IP
func generateConfigs(tpl integration.Config, kep *v1.Endpoints) []integration.Config
```

Note: we can also use `<PREFIX><UID>/<IP>` but it's more human friendly this way `<PREFIX><NAMESPACE>/<NAME>/<IP>` as we have the uniqueness guaranteed in both options.

Reply:
Got it, I would have expected the UIDs to be different for each endpoints. But I agree it's more human friendly.